nssa/src/state.rs

```diff
@@ -2221,6 +2223,1547 @@ pub mod tests {
         ));
     }
 
+    //TODO: repeated code needs to be cleaned up
+    //from token.rs (also repeated in amm.rs)
+    const TOKEN_DEFINITION_TYPE: u8 = 0;
+    const TOKEN_DEFINITION_DATA_SIZE: usize = 23;
+
+    const TOKEN_HOLDING_TYPE: u8 = 1;
+    const TOKEN_HOLDING_DATA_SIZE: usize = 49;
+
+    struct TokenDefinition {
+        account_type: u8,
+        name: [u8; 6],
+        total_supply: u128,
+    }
+
+    struct TokenHolding {
+        account_type: u8,
+        definition_id: AccountId,
+        balance: u128,
+    }
+    impl TokenDefinition {
+        fn into_data(self) -> Data {
+            let mut bytes = [0; TOKEN_DEFINITION_DATA_SIZE];
+            bytes[0] = self.account_type;
+            bytes[1..7].copy_from_slice(&self.name);
+            bytes[7..].copy_from_slice(&self.total_supply.to_le_bytes());
+            bytes
+                .to_vec()
+                .try_into()
+                .expect("23 bytes should fit into Data")
+        }
+    }
+
+    impl TokenHolding {
+        fn new(definition_id: &AccountId) -> Self {
+            Self {
+                account_type: TOKEN_HOLDING_TYPE,
+                definition_id: definition_id.clone(),
+                balance: 0,
+            }
+        }
+
+        fn parse(data: &[u8]) -> Option<Self> {
+            if data.len() != TOKEN_HOLDING_DATA_SIZE || data[0] != TOKEN_HOLDING_TYPE {
+                None
+            } else {
+                let account_type = data[0];
+                let definition_id = AccountId::new(data[1..33].try_into().unwrap());
+                let balance = u128::from_le_bytes(data[33..].try_into().unwrap());
+                Some(Self {
+                    definition_id,
+                    balance,
+                    account_type,
+                })
+            }
+        }
+
+        fn into_data(self) -> Data {
+            let mut bytes = [0; TOKEN_HOLDING_DATA_SIZE];
+            bytes[0] = self.account_type;
+            bytes[1..33].copy_from_slice(&self.definition_id.to_bytes());
+            bytes[33..].copy_from_slice(&self.balance.to_le_bytes());
+            bytes
+                .to_vec()
+                .try_into()
+                .expect("33 bytes should fit into Data")
+        }
+    }
+
+    //TODO repeated code should ultimately be removed;
```

Comment:
So as I understand there should be an amm library crate which will be included in risc0 program and these tests?